### PR TITLE
Add Item overlay to show when monsters are weak enough to finish off

### PR DESCRIPTION
**Pull request by Berry-95**
Fixes: #784
Current behaviour shows the required item to finish an NPC off above it's head when it's health is below the required threshold.
Works for Rockslugs, Gargoyles and Desert Lizards. (Let me know if I forgot any others).

### CODE REVIEW

File: runelite-client/src/main/java/net/runelite/client/plugins/slayer/SlayerConfig.java [hunk 1]
```diff
@@ -99,6 +99,16 @@ default Color getTargetColor()
 		return Color.RED;
 	}
 
+	@ConfigItem(
+			position = 7,
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/slayer/SlayerConfig.java]: Fix the indentation (just import XML here: https://github.com/runelite/runelite/wiki/Code-Conventions to IntelliJ)

File: runelite-client/src/main/java/net/runelite/client/plugins/slayer/SlayerConfig.java [hunk 2]
```diff
@@ -99,6 +99,16 @@ default Color getTargetColor()
...
+			keyName = "weaknessPrompt",
+			name = "Show Monster Weakness",
+			description = "Show an overlay on a monster when it is weak enough to finish off (Only Lizards, Gargoyles & Rockslugs)"
+	)
+	default boolean weaknessPrompt()
+	{
+		return true;
+	}
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/slayer/SlayerConfig.java]: add empty line below

File: runelite-client/src/main/java/net/runelite/client/plugins/slayer/TargetWeaknessOverlay.java [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+public class TargetWeaknessOverlay extends Overlay
+{
+	private final Client client;
+	private final SlayerConfig config;
+	private final SlayerPlugin plugin;
+	private final ItemManager itemManager;
+
+	@Inject
+	TargetWeaknessOverlay(Client client, SlayerConfig config, SlayerPlugin plugin, ItemManager itemManager)
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/slayer/TargetWeaknessOverlay.java]: Make this ctor private

File: runelite-client/src/main/java/net/runelite/client/plugins/slayer/TargetWeaknessOverlay.java [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+import net.runelite.api.Perspective;
+import net.runelite.api.Point;
+import net.runelite.api.coords.LocalPoint;
+import net.runelite.client.game.ItemManager;
+import net.runelite.client.ui.overlay.Overlay;
+import net.runelite.client.ui.overlay.OverlayLayer;
+import net.runelite.client.ui.overlay.OverlayPosition;
+import net.runelite.client.ui.overlay.OverlayUtil;
+
+public class TargetWeaknessOverlay extends Overlay
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/slayer/TargetWeaknessOverlay.java]: Remove public modifier

**Comment by deathbeam** [comment]
I would also probably make this enum/reuse existing slayer enum, and just add itemId/hp for specific npcs

File: runelite-client/src/main/java/net/runelite/client/plugins/slayer/SlayerConfig.java [hunk 5]
```diff
@@ -99,6 +99,14 @@ default Color getTargetColor()
 		return Color.RED;
 	}
 
+	@ConfigItem(
+		position = 7,
+		keyName = "weaknessPrompt",
+		name = "Show Monster Weakness",
+		description = "Show an overlay on a monster when it is weak enough to finish off (Only Lizards, Gargoyles & Rockslugs)"
+	)
+	default boolean weaknessPrompt() { return true;	}
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/slayer/SlayerConfig.java]: empty lines

File: runelite-client/src/main/java/net/runelite/client/plugins/slayer/TargetWeaknessOverlay.java [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+		}
+
+		Task curTask = Task.getTask(plugin.getTaskName());
+		if (curTask == null)
+		{
+			return null;
+		}
+
+		float threshold = 0f;
+		switch (curTask)
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/slayer/TargetWeaknessOverlay.java]: You should probably save the threshold in `Task` too

File: runelite-client/src/main/java/net/runelite/client/plugins/slayer/TargetWeaknessOverlay.java [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+		}
+		return null;
+	}
+
+	private void renderTargetItem(Graphics2D graphics, NPC actor, Task curTask)
+	{
+		BufferedImage image = null;
+		switch (curTask)
+		{
+			case DESERT_LIZARDS:	image = itemManager.getImage(ItemID.ICE_COOLER);	break;
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/slayer/TargetWeaknessOverlay.java]: And this too (e.g the ItemID)

File: runelite-client/src/main/java/net/runelite/client/plugins/slayer/SlayerConfig.java [hunk 8]
```diff
@@ -105,7 +105,8 @@ default Color getTargetColor()
 		name = "Show Monster Weakness",
 		description = "Show an overlay on a monster when it is weak enough to finish off (Only Lizards, Gargoyles & Rockslugs)"
 	)
-	default boolean weaknessPrompt() { return true;	}
+	default boolean weaknessPrompt()
+	{ return true; }
```
[the remark this answers is not in the document]
Comment (Berry-95) [reply runelite-client/src/main/java/net/runelite/client/plugins/slayer/SlayerConfig.java]: Ahhh I see, not entirely obvious as the rest of the config has the wrong checkstyle but I got it 
👍

[the remark this answers is not in the document]
Comment (deathbeam) [reply runelite-client/src/main/java/net/runelite/client/plugins/slayer/SlayerConfig.java]: It doesnt, you just have it collapsed in your IDE

[the remark this answers is not in the document]
Comment (Berry-95) [reply runelite-client/src/main/java/net/runelite/client/plugins/slayer/SlayerConfig.java]: Well now I just feel silly 😆

File: runelite-client/src/main/java/net/runelite/client/plugins/slayer/TargetWeaknessOverlay.java [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+			return null;
+		}
+
+		Task curTask = Task.getTask(plugin.getTaskName());
+		if (curTask == null || curTask.getWeaknessThreshold() < 0)
+		{
+			return null;
+		}
+
+		float threshold = curTask.getWeaknessThreshold();
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/slayer/TargetWeaknessOverlay.java]: you should probably also retrieve the icon here too, and if there is no weakness icon (e.g -1) do early return

File: runelite-client/src/main/java/net/runelite/client/plugins/slayer/TargetWeaknessOverlay.java [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+		}
+
+		Task curTask = Task.getTask(plugin.getTaskName());
+		if (curTask == null || curTask.getWeaknessThreshold() < 0 || curTask.getWeaknessItem() < 0)
+		{
+			return null;
+		}
+
+		float threshold = curTask.getWeaknessThreshold();
+		BufferedImage image = itemManager.getImage(curTask.getWeaknessItem());
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/slayer/TargetWeaknessOverlay.java]: you should first check for weaknessItem (as -1 will not work) and do the early return

Comment (deathbeam) [reply runelite-client/src/main/java/net/runelite/client/plugins/slayer/TargetWeaknessOverlay.java]: Oh actually, you did that at top. you could've reused the variable, but, I guess this is fine

File: runelite-client/src/main/java/net/runelite/client/plugins/slayer/TargetWeaknessOverlay.java [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+			}
+		}
+		return null;
+	}
+
+
+	/**
+	 * Based on how health is calculated in OpponentInfoOverlay
+	 */
+	private int calculateHealth(NPC target)
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/slayer/TargetWeaknessOverlay.java]: Is this really needed here? probably not, it does not need to be that accurate

Comment (Berry-95) [reply runelite-client/src/main/java/net/runelite/client/plugins/slayer/TargetWeaknessOverlay.java]: It's a way of dealing with superiors and alternatives having different max health and therefore a different health ratio threshold, even though the numeric threshold is the same.
Although alternative suggestions are welcome, I could for example compare ids/level and hardcode the threshold for each one.

Comment (deathbeam) [reply runelite-client/src/main/java/net/runelite/client/plugins/slayer/TargetWeaknessOverlay.java]: Then all you need is hp and divide by ratio. you dont need the rest of the logic

**Comment by Hampo** [comment]
Plugin description doesn't mention zygomites.

**Pull request merged**